### ISSUE TITLE
`update` fails to add a new record type

### ISSUE THREAD

**Issue by aryairani** [unisonweb/unison#4425]
I created https://github.com/unisonweb/unison/tree/fix/update-new-record with this failing transcript to kick things off.

```unison
unique type Foo = { bar : Nat }
```

```ucm

  I found and typechecked these definitions in scratch.u. If you
  do an `add` or `update`, here's how your codebase would
  change:
  
    ⍟ These new definitions are ok to `add`:
    
      unique type Foo
      Foo.bar        : Foo -> Nat
      Foo.bar.modify : (Nat ->{g} Nat) -> Foo ->{g} Foo
      Foo.bar.set    : Nat -> Foo -> Foo

```
This shouldn't be an error.

```ucm
.> update

  Okay, I'm searching the branch for code that needs to be
  updated...

  That's done. Now I'm making sure everything typechecks...

  Everything typechecks, so I'm saving the results...

  I couldn't complete the update because I couldn't find 0
  constructor(s) for Foo where I expected to. I found: []
  
  You can use `view Foo` and
  `alias.term <hash> Foo.<ConstructorName>` to give names to
  each constructor, and then try again.

.> view Foo

  ⚠️
  
  The following names were not found in the codebase. Check your spelling.
    Foo

```

**Comment by ceedubs** [comment]
I keep running into this. Do we know a workaround until it is addressed?

This part of the error message is bewildering to me:

> I couldn't find 0 constructor(s) for Foo where I expected to. I found: []

If you found an empty list then you could in fact find 0 constructors! But also that's weird because you just created the type with a single constructor.

**Comment by aryairani** [comment]
@ceedubs A workaround for now is to use `add Foo` to add `Foo` before using `update` to add/update any of the rest of the definitions in the file.

**Comment by aryairani** [comment]
It looks like `typecheckedUnisonFileToBranchUpdates` tries to delete constructors for any types that are in the `TypecheckedUnisonFile`, even if the type didn't exist in the namespace previously.

Also `findCtorNames` returns an error if queried with a type name that didn't exist.

`findCtorNames` doesn't have access to the type names though, so it could either silently return `[]`, or else `typecheckedUnisonFileToBranchUpdates` should not call `findCtorNames` for a type that doesn't exist in the branch.

~~I'm thinking `makeAnnihilateTypeName` should return an indication of whether it did anything; if it didn't, then let `deleteStuff = []`.~~ This doesn't make sense.

**Comment by aryairani** [comment]
I added trace statements until I started working; no obvious logic change yet.

**Comment by aryairani** [comment]
I think I have a fix for this; it was a signed int (map size) vs unsigned int (constructor id) issue, combined with laziness which allowed it to short circuit with failure instead of hanging while it counted up to max_int, which would've been more noticeable to fix.

**Comment by aryairani** [comment]
> I keep running into this. Do we know a workaround until it is addressed?

I put up a PR for the fix.

> This part of the error message is bewildering to me:
> 
> > I couldn't find 0 constructor(s) for Foo where I expected to. I found: []
> 
> If you found an empty list then you could in fact find 0 constructors! But also that's weird because you just created the type with a single constructor.

Yeah this was super confusing. It really only needed 0 but was failing for a different reason.

**Comment by aryairani** [comment]
Closed by #4434
